src/components/magic-wrapper.js

```diff
@@ -0,0 +1,223 @@
+
+const Deck = styled.div(() => ({
+  position: 'absolute',
+  top: 0,
+  left: 0,
+  width: '100%',
+  height: '100%',
+}));
+
+class Context extends React.Component {
+  static contextTypes = {
+    contentHeight: PropTypes.number,
+    contentWidth: PropTypes.number,
+    export: PropTypes.bool,
+    overview: PropTypes.bool,
+    print: PropTypes.bool,
+    store: PropTypes.object,
+    styles: PropTypes.object,
+  };
+  static childContextTypes = {
+    contentHeight: PropTypes.number,
+    contentWidth: PropTypes.number,
+    export: PropTypes.bool,
+    overview: PropTypes.bool,
+    print: PropTypes.bool,
+    store: PropTypes.object,
+    styles: PropTypes.object,
+  };
+  static propTypes = {
+    children: PropTypes.node,
+    context: PropTypes.object,
+  };
+  getChildContext() {
+    return {
+      contentHeight: this.props.context.contentHeight,
+      contentWidth: this.props.context.contentWidth,
+      export: this.props.context.export,
+      overview: this.props.context.overview,
+      print: this.props.context.print,
+      store: this.props.context.store,
+      styles: this.props.context.styles,
+    };
+  }
+  render() {
+    return this.props.children;
+  }
+}
+
+export default class MagicText extends React.Component {
+  static contextTypes = {
+    contentHeight: PropTypes.number,
+    contentWidth: PropTypes.number,
+    export: PropTypes.bool,
+    overview: PropTypes.bool,
+    print: PropTypes.bool,
+    store: PropTypes.object,
+    styles: PropTypes.object,
+  };
+  static propTypes = {
+    children: PropTypes.node,
+    exitSubscription: PropTypes.func,
+    magicIndex: PropTypes.number,
+    presenter: PropTypes.bool
+  };
+  constructor(props) {
+    super(props);
+    this.container = null;
+    this.styleMap = {};
+    this.lastPortalMap = {};
+    this.portalMap = {};
+    this.diffs = {};
+    this.lastDiffs = null;
+    this.state = {
+      renderedChildren: props.children,
+    };
+  }
+
+  componentDidMount() {
+    this.mounted = true;
+    this.portal = document.getElementById('portal');
+    if (!this.props.presenter) {
+      this.container.animate([{ opacity: 0 }, { opacity: 1 }], {
+        duration: 500,
+        easing: 'ease-in',
+      });
+      this.props.exitSubscription(() => {
+        this.container.animate([{ opacity: 1 }, { opacity: 0 }], {
+          duration: 500,
+          easing: 'ease-in',
+        });
+      });
+    }
+    if (!this.portal) {
+      this.portal = this.makePortal();
+    }
+    ReactDOM.render(
+      <Context context={this.context}>
+        <Deck>{this.props.children}</Deck>
+      </Context>,
+      this.portal,
+      () => {
+        this.timeout = setTimeout(() => {
+          updateChildren(this.container.childNodes[0]);
+          updateChildren(this.portal.childNodes[0].childNodes[0]);
+          buildStyleMap(
+            this.portalMap,
+            this.portal.childNodes[0].childNodes[0]
```

Comment:
Will the first element of these arrays always exist?

Reply:
Yes, but i'll code defensively against it anyway